### PR TITLE
[C2R 1] Update Lightning to support the new Catalog2Registration with -driver

### PR DESCRIPTION
**Pull request by joelverhagen**
Address https://github.com/nuget/nugetgallery/issues/7740
Depends on https://github.com/NuGet/NuGet.Services.Metadata/pull/727

### CODE REVIEW

File: src/Ng/Jobs/LightningJob.cs [hunk 1]
```diff
@@ -492,36 +516,257 @@ private static Task UpdateCursorFileAsync(string cursorFileName, int startIndex,
...
-
-            if (_storageFactories.SemVer2StorageFactory != null)
+            var builder = new StringBuilder();
+            builder.Append("DefaultEndpointsProtocol=https;");
+            builder.AppendFormat("AccountName={0};", _arguments.GetOrThrow<string>(accountNameArgument));
+            builder.AppendFormat("AccountKey={0};", _arguments.GetOrThrow<string>(accountKeyArgument));
+            builder.AppendFormat("EndpointSuffix={0}", _arguments.GetOrDefault(endpointSuffixArgument, "core.windows.net"));
+
+            var connectionString = builder.ToString();
+            if(currentConnectionString != null && currentConnectionString != connectionString)
```
Comment (loic-sharma) [inline src/Ng/Jobs/LightningJob.cs]: Unforgivable nit: 
```suggestion
            if (currentConnectionString != null && currentConnectionString != connectionString)
```

File: src/Ng/Jobs/LightningJob.cs [hunk 2]
```diff
@@ -492,36 +516,257 @@ private static Task UpdateCursorFileAsync(string cursorFileName, int startIndex,
...
+                // Update the hives.
+                await _updater.UpdateAsync(_packageId, entries, entryToLeaf);
+
+                _urls.Clear();
             }
         }
+
+        private interface IStrike
+        {
+            Task ProcessCatalogLeafUrl(string url);
```
Comment (loic-sharma) [inline src/Ng/Jobs/LightningJob.cs]: Nit:

```suggestion
            Task ProcessCatalogLeafUrlAsync(string url);
```

File: src/Ng/Jobs/LightningJob.cs [hunk 3]
```diff
@@ -326,48 +358,38 @@ private async Task PrepareAsync()
...
+                    var optionalArguments = new StringBuilder();
+                    AppendOptionalArgument(optionalArguments, Arguments.ContentIsFlatContainer);
+                    AppendOptionalArgument(optionalArguments, Arguments.FlatContainerName);
+                    AppendOptionalArgument(optionalArguments, Arguments.StorageSuffix);
+                    AppendOptionalArgument(optionalArguments, Arguments.AllIconsInFlatContainer);
+                    AppendOptionalArgument(optionalArguments, Arguments.Driver);
+                    AppendOptionalArgument(optionalArguments, Arguments.Verbose);
+
+                    commandStreamContents = commandStreamContents
+                            .Replace($"[{optionalArgumentsTemplate}]", optionalArguments.ToString());
```
Comment (agr) [inline src/Ng/Jobs/LightningJob.cs]: extra tab?

File: src/Ng/Jobs/LightningJob.cs [hunk 4]
```diff
@@ -492,36 +516,257 @@ private static Task UpdateCursorFileAsync(string cursorFileName, int startIndex,
...
+                        }
+                    });
+
+                // Build the input to hive updaters.
+                var entries = new List<CatalogCommitItem>();
+                var entryToLeaf = new Dictionary<CatalogCommitItem, PackageDetailsCatalogLeaf>(
+                    ReferenceEqualityComparer<CatalogCommitItem>.Default);
+                foreach (var leaf in leaves)
+                {
+                    if (!(leaf.IsPackageDelete() ^ leaf.IsPackageDetails()))
```
Comment (agr) [inline src/Ng/Jobs/LightningJob.cs]: Is that a complicated way to write `leaf.IsPackageDelete() == leaf.IsPackageDetails()`?

Comment (joelverhagen) [reply src/Ng/Jobs/LightningJob.cs]: 😅 yes. Your's is much better!

_"but mine uses a carrot!"_

File: src/Ng/Jobs/LightningJob.cs (hunk 3 above)
Comment (joelverhagen) [reply src/Ng/Jobs/LightningJob.cs]: 🔍

**Pull request merged**